src/staticdata_utils.c

```diff
@@ -1162,21 +1162,35 @@ static void jl_insert_backedges(jl_array_t *edges, jl_array_t *ext_targets, jl_a
             }
         }
         // then enable any methods associated with it
-        void *ci = ptrhash_get(&visited, (void*)caller);
+        void *ci = ptrhash_get(&cis_pending_validation, (void*)caller);
         //assert(ci != HT_NOTFOUND);
         if (ci != HT_NOTFOUND) {
-            // have some new external code to use
+            // Update any external CIs and add them to the cache.
             assert(jl_is_code_instance(ci));
             jl_code_instance_t *codeinst = (jl_code_instance_t*)ci;
             assert(codeinst->min_world == minworld && jl_atomic_load_relaxed(&codeinst->inferred) );
             codeinst->max_world = maxvalid;
-            if (jl_rettype_inferred(caller, minworld, maxvalid) == jl_nothing) {
-                jl_mi_cache_insert(caller, codeinst);
+
+            if (jl_rettype_inferred(codeinst->def, minworld, maxvalid) != jl_nothing) {
+                // We already got a code instance for this world age range from somewhere else - we don't need
+                // this one.
+                continue;
+            }
+
+            jl_mi_cache_insert(codeinst->def, codeinst);
+        } else {
+            // Likely internal. Find the CI already in the cache hierarchy.
+            for (jl_code_instance_t *codeinst = caller->cache; codeinst; codeinst=codeinst->next) {
+                if (codeinst->min_world != minworld)
+                    continue;
+                if (codeinst->max_world != WORLD_AGE_REVALIDATION_SENTINEL)
+                    continue;
+                codeinst->max_world = maxvalid;
             }
```

Comment:
This visitor seems slightly uncomfortable to me. There seems an assumption to me that all CI are handled by `jl_insert_backedges` as well, which is not true (some don't have any edges at all)

Reply:
If it doesn't have any edges (and it is internal), the serializer doesn't set the `max_world` to `WORLD_AGE_REVALIDATION_SENTINEL`, and no further processing is required.

Reply:
What if it wasn't internal? Clearly I don't remember well how this part of the code is currently implemented, as it has changed too many times

Reply:
If it's not internal, it's in the `next_ext_cis` list and will not got through this code path.

Reply:
I still don't really like this iterator, though maybe it is unavoidable to bring it back. It seems to imply that it is forbidden for multiple codeinst to exist for roughly the same set of conditions (as they might have different edge constraints), and that isn't really something we want to disallow. And seems like it implies that deserialization did a bad job at tracking the content it loaded, which seems awkward and unfortunate.

Reply:
I don't think I understand what you're saying. There's nothing preventing multiple code instances from being in this cache. Are you saying that the edges are really a property of the CodeInstances rather than the MethodInstances? In that case, I think I would agree with that, but that's not really how things are set up right now. I really think this is a string improvement over what we used to do, which would result in essentially the same end state, but much more convoluted. If we want to do even better than this, I'm all in favor of course.

Reply:
Maybe it is also just the comment about "Likely internal". Should that say simply "Internal"? (and optionally assert that with a pkgimage id lookup via the address)

Yeah, agreed that edges right now are coalesced into the MethodInstance even though they belong to the CodeInstance. It is one of the things that https://github.com/JuliaLang/julia/pull/52415 is intended to start working towards fixing, as it better stores (some of) these edges, so they are not lost into the Method.roots table instead (as happens now)

Reply:
I think it's probably possible for something to be neither external nor internal depending on what external absints are doing, which is the case I was hedging here. I can change the comment if you prefer though.

Reply:
That seems like it would be an error to end up here though in any other case, right? Since this is validating edges, except those edges will validate incorrectly for any absint. Particularly any absint that doesn't do vanilla inference and/or that uses overlay tables, this code will be wrong for them and invalid to reach here with those.

Reply:
Yes, it shouldn't end up here (though I think it would probably be reasonably safe). We'll want to think about what the proper thing to do for external absints is, but for now, I think it would be fine.